front: adds opId into PathOperationalPoint

jacomyal opened the pull request:
This commit aims at clarifying the `OperationalPoint` types (there are two identical types) that in practice have IDs that are not actual OP IDs, because the Manchette needs unique IDs, basically. It will help fetching track occupation data, when adding a track-occupancy diagram into the space-time chart (see [related ticket](https://github.com/osrd-project/osrd-confidential/issues/866)).

The solution is to also store the actual OP ID when it's possible, and get types with a slightly better naming, **`PathOperationalPoint`**, to highlight that these represent operational points, within the context of a given path.

#### Details:

- Replaces type `OperationalPoint` from both `operationalStudies/types` and `stdcm/types`, with **`EditoastPathOperationalPoint`** (exact data from Editoast) and **`PathOperationalPoint`** (with unique `waypointId` and optional `opId` instead of `id`) in `simulationResult/types`
- Switches to the relevant types in both STDCM and Operational Studies
- Stores the proper `opId` in the two places where the front generates unique IDs for the OPs for the Manchette
- Adapts `upsertMapWaypointsInOperationalPoints` so that it can both handles `PathOperationalPoint` or `EditoastPathOperationalPoint`
- Adds some comment about some fishy code in STDCM

jacomyal committed:
front: adds opId into PathOperationalPoint

This commit aims at clarifying the OperationalPoint types (there are two
identical types) that in practice have IDs that are not actual OP IDs,
because the Manchette needs unique IDs, basically.

The solution is to also store the actual OP ID when it's possible, and
get types with a slightly better naming, PathOperationalPoint, to
highlight that these represent operational points, within the context of
a given path.

Details:
- Replaces type OperationalPoint from both operationalStudies/types and
  stdcm/types, by EditoastPathOperationalPoint (which is the data from
  Editoast) and PathOperationalPoint (with a unique waypointId, and an
  optional opId)
- Switches to the relevant types in both STDCM and Operational Studies
- Stored the proper opId in the two places where the front generates
  unique IDs for the OPs for the Manchette
- Adapts upsertMapWaypointsInOperationalPoints so that it can handle
  both PathOperationalPoints and EditoastPathOperationalPoints

Signed-off-by: Alexis Jacomy <alexis.jacomy@gmail.com>

SharglutDev commented on front/src/applications/stdcm/utils/stdcmComputeChartData.ts:
```diff
@@ -17,7 +18,7 @@ const computeChartData = (
   const formattedPathProperties = preparePathPropertiesData(
     simulation.electrical_profiles,
-    pathProperties,
+    pick(pathProperties, 'geometry', 'zones'),
```
Does this change concerns the related issue ? Maybe it could be done in a separate commit.

Nevertheless, this could use some comments of why we don't need the other props (don't we need to prepare slopes and curves for the speed space chart ?)

jacomyal replied:
So, here's the thing about this bit:
- `preparePathPropertiesData` requires a second argument of type **`PathProperties`**
- `pathProperties` is a **`StdcmPathProperties`**

I don't even know if it is on purpose or not. But TypeScript is ok with that, because all keys from `PathProperties` are optional. As `StdcmPathProperties` shares two keys with `PathProperties` (`geometry` and `zones`), I thought adding this `pick(...)` would clarify what happens, or at least bring attention to this part. What do you think about it?

SharglutDev replied:
Oh right didn't notice that at first read.

So to be sure to understand, when debugging for your opId issue you noticed this oddity and you made this change ?
I totally think we can benefit from this change but still think it should be put in a different commit where you could add a description :)

jacomyal replied:
Agreed, thanks. I'll submit a new PR then.

SharglutDev replied:
It's perfectly fine to me if you put the commit in this PR as it doesn't change any behavior :)

jacomyal replied:
I just pushed a separated commit. Also, instead of calling `pick`, I just added some more detailed comment about why the code is fishy.

jacomyal committed:
front: adds some TODO about some fishy code

The comment says it all, but basically, some data of some shape is given
to a function that requires another shape, and it might hide some
potential bug.

Signed-off-by: Alexis Jacomy <alexis.jacomy@gmail.com>

emersion commented on front/src/applications/operationalStudies/helpers/upsertMapWaypointsInOperationalPoints.ts:
```diff
@@ -57,7 +63,7 @@ export const upsertMapWaypointsInOperationalPoints = (
           position: positionOnPath,
           weight: HIGHEST_PRIORITY_WEIGHT,
-        };
+        } as Point;
```
I prefer to avoid casts. Here `Point` can be any type with extra properties, and these extra properties might be missing from the object we generate. It seems like typing is incorrect?

Why do we need a generic here in the first place?

jacomyal replied:
I confess I didn't like this cast either :sweat_smile: I had to add it, because otherwise TypeScript wasn't happy with a potential `NonNullable<PathProperties['operational_points']>[number]` with a `opId` property.

More precisely, here is the TypeScript error when I use `const formattedStep: Point = {...};` instead:

```
TS2322: Type
{
  id: string;
  opId: null;
  extensions: {
    identifier: {
      name: string;
      uic: number;
    };
  };
  part: {
    track: string;
    position: number;
  };
  position: number;
  weight: number;
}
is not assignable to type Point:
{
  id: string; 
  opId: null; 
  extensions: {
    identifier: {
      name: string; 
      uic: number; 
    }; 
  }; 
  part: {
    track: string; 
    position: number; 
  }; 
  position: number; 
  weight: number; 
}
is assignable to the constraint of type Point, but Point could be instantiated with a different subtype of constraint
{
  extensions?: OperationalPointExtensions | undefined;
  id: string;
  part: OperationalPointPart;
  position: number;
  weight: number | null;
} | PathOperationalPoint
```

@emersion How would you solve that, in that context?

SharglutDev replied:
Would it be possible to just do : 
```
export function upsertMapWaypointsInOperationalPoints(
  path: TrainSchedule['path'],
  pathItemsPositions: PathfindingResultSuccess['path_item_positions'],
  operationalPoints: (NonNullable<PathProperties['operational_points']>[number] | PathOperationalPoint)[],
  t: TFunction<'operational-studies'>
): (NonNullable<PathProperties['operational_points']>[number] | PathOperationalPoint)[] {
```

And then 
```
const formattedStep: PathOperationalPoint = {
```

emersion commented on front/src/modules/simulationResult/types.ts:
```diff
@@ -18,6 +17,10 @@ import type { PacedTrainWithDetails } from 'modules/trainschedule/components/Tim
 import type { OccurrenceId, TimetableItem, TrainScheduleId } from 'reducers/osrdconf/types';
 import type { ArrayElement } from 'utils/types';
 
+export type PathOperationalPoint = NonNullable<PathProperties['operational_points']>[number] & {
+  opId: string | null;
+};
```
I think defining a new type is a great idea! However, I find it confusing to silently re-purpose the `id` field.

In the original `PathProperties['operational_points']` type, the `id` field is an OP ID. In `PathOperationalPoint`, the `id` field is a waypoint ID.

I would suggest to either:

- Leave the `id` field from the original type with the same semantics (OP ID), drop `opId`, and add a new separate field for the waypoint ID.
- Omit the `id` field from `PathOperationalPoint`, keep `opId`, and define a new, different separate field for the waypoint ID.

If these options are too much work, we could also add a doc comment explaining this subtlety, but I'd still find this surprising and inconsistent with editoast.

jacomyal replied:
I agree, I went that direction because of the impact. I'll check again how much work it would require to store the custom waypoint ID in another field.

emersion reviewed (commented):
> Removes type OperationalPoint from both operationalStudies/types and stdcm/types

Thanks so much for doing this! These old types were confusing.

I left two comments above.

jacomyal commented:
@emersion @SharglutDev I just pushed a new version of the PR.

- I improved the new type, and it now has the unique `waypointId` and an optional `opId` (it was less impactful than I thought)
- I added some comments to clarify the difference between `PathOperationalPoint` and `EditoastPathOperationalPoint`
- I switched from the generic to some polymorphism for `upsertMapWaypointsInOperationalPoints`, so that it provides data with the proper shape

emersion reviewed (approved):
Thanks, this looks good to me!

jacomyal commented:
Oh, one last thing I forgot to mention @emersion @SharglutDev  :sweat_smile::

This commit breaks the compatibility of the data cached in the local storage:

- We write in the local storage [here](https://github.com/OpenRailAssociation/osrd/blob/e62c011f9bc7cb2ee8aa91d7813866c1c96dec47/front/src/modules/simulationResult/components/SpaceTimeChart/WaypointsPanel.tsx#L76-L79)
- We read from it [here](https://github.com/OpenRailAssociation/osrd/blob/e62c011f9bc7cb2ee8aa91d7813866c1c96dec47/front/src/modules/simulationResult/components/SpaceTimeChart/useGetProjectedTrainOperationalPoints.ts#L88-L90)

 Has this already occurred in the past? Should I add some code to validate the data shape? Or should I add some version-tag in the local storage key?

SharglutDev commented:
With the approval of @clarani, seems like we don't have any user having a study on osrd right now so it shouldn't impact anyone "important".

(But the version thing could be an enhancement for later maybe ?)

emersion commented:
Oh. That's a good point.

Our policy so far: we absolutely want to avoid crashing the webapp in production, but it's fine if developers need to clear the local storage for their local development setup. It's fine if data gets wiped in production. (I don't think this is a good long-term policy, but that's a different discussion.)

We've been using redux-persist for most of our local storage needs, and we clear the redux data each version upgrade on deployed environments. We don't have version info on development setups, so we don't clear it there. This all fits the policy.

But when using `localStorage` directly without redux, the old data persists across version upgrades in production. I suppose the new code will crash or misbehave with old data because it'll be missing a `waypointId` field?

If so, I think we should indeed catch this case, either by throwing away old data, by using a different local storage key, or by converting to the new type when loading from local storage.

(Side note, I don't think we should persist everything we persist here: we don't need the full OP metadata.)

SharglutDev commented:
Oh thought we would just lost the previous saved manchette but if it's breaking the app we should handle the case yes

jacomyal committed:
front: breaks some localStorage key on purpose

This commit comes after this conversation:
https://github.com/OpenRailAssociation/osrd/pull/11838#issuecomment-2898134465

Details:
- Moves three hardcoded places where a LocalStorage key is generated in
  the simulation results module to a new helper, getWaypointsLocalStorageKey
- The new key is also different from the previous one (it has a prefix),
  because of the type changes from the previous commit

Signed-off-by: Alexis Jacomy <alexis.jacomy@gmail.com>

jacomyal commented:
OK, I added a commit that changes the key (it adds the new type name as prefix). It also moves the key generation part to a new helper.

emersion reviewed (approved):
Nice! Yeah adding a prefix is a good idea regardless.

SharglutDev reviewed (approved):
Thank you for the PR
